docs/sections/contributor_guide/code_quality.rst

```diff
@@ -1,5 +1,36 @@
-************
 Code Quality
-************
+============
 
-TBD
+.. include:: links.rst
+
+In an active development shell, the following ``make`` targets are available and act on all ``.py`` files under ``src/``:
+
+.. list-table::
+   :widths: 15 85
+   :header-rows: 1
+
+   * - Command
+     - Description
+   * - ``make format``
+     - Formats code with `black`_, imports with `isort`_, docstrings with `docformatter`_, and ``.jsonschema`` documents with `jq`_
+   * - ``make lint``
+     - Lints code with `pylint`_
+   * - ``make typecheck``
+     - Typechecks code with `mypy`_
+   * - ``make unittest``
+     - Runs unit tests and reports coverage with `pytest`_ and `coverage`_
+   * - ``make test``
+     - Equivalent to ``make lint && make typecheck && make unittest``, plus checks defined CLI scripts.
+
+Configuration for these tools is provided by the file ``src/pyproject.toml``.
+
+The order of the targets above is intentional:
+
+   * ``make format`` will complain about certain kinds of syntax errors that would cause all the remaining code-quality tools to fail (and may change line numbers reported by other tools, if it ran after them).
+   * ``make lint`` provides a good first check for obvious errors and anti-patterns in the code.
+   * ``make typecheck`` offers a more nuanced look at interfaces between functions, methods, etc. and may spot issues missed by the linter.
+   * ``make unittest`` provides higher-level semantic-correctness checks once code syntax and typing is deemed correct.
+
+All the above tests are executed by the CI system when code is merged to specific git branches, and again when a conda package is built for release. To ensure that these processes succeed, be sure to run all the tests in a development shell before opening a pull request, and throughout the PR's lifecycle as subsequent changes are made. CI will reject unformatted code, so also run ``make format`` and commit any changes it makes. A useful development idiom is to periodically run ``make format && make test`` to perform a full code-quality sweep through the code.
```

Comment:
Speaking of wordy, I revised this paragraph and made it more concise. Please let me know if you think it still needs work.